controllers/metal3.io/baremetalhost_controller.go

```diff
@@ -757,12 +757,18 @@ func (r *BareMetalHostReconciler) actionPreparing(prov provisioner.Provisioner,
 	}
 
 	prepareData := provisioner.PrepareData{
-		RAIDConfig:      newStatus.Provisioning.RAID.DeepCopy(),
-		RootDeviceHints: newStatus.Provisioning.RootDeviceHints.DeepCopy(),
-		FirmwareConfig:  newStatus.Provisioning.Firmware.DeepCopy(),
+		TargetRAIDConfig:  newStatus.Provisioning.RAID.DeepCopy(),
+		ExistedRAIDConfig: info.host.Status.Provisioning.RAID,
+		RootDeviceHints:   newStatus.Provisioning.RootDeviceHints.DeepCopy(),
+		FirmwareConfig:    newStatus.Provisioning.Firmware.DeepCopy(),
+	}
+	// When manual cleaning fails, we think that the existed RAID configuration
+	// is invalid and needs to be reconfigured.
+	if info.host.Status.ErrorType == metal3v1alpha1.PreparationError {
```

Comment:
@zaneb 
> How can we avoid assuming that there is no RAID configured on the host after an error occurs in Preparing?

We seem to be able to avoid this problem in this way 。

Reply:
@Hellcatlk in theory the `PreparationError` could be raised also for other reasons than an invalid RAID (ie potentially an invalid BIOS settings). Have you maybe verified functionally this case? If the original error doesn't not depend on the RAID settings I guess that it would be safe to re-apply the same cleaning steps configured for RAID by Ironic, but it'd be nice to confirm it to be on the safe side (instead, if the error was due the invalid RAID, the retry mechanism triggered should give enough room to the user to fix it as required).

Reply:
@andfasano Sorry I miss this.

> in theory the PreparationError could be raised also for other reasons than an invalid RAID (ie potentially an invalid BIOS settings). Have you maybe verified functionally this case?

Yes, no matter what causes the preparingError, it will reconfigure RAID. 

The following two places will configure the same RAID as `Status.Provisioning.RAID` when the user setting `Spec.RAID` is empty. 

https://github.com/metal3-io/baremetal-operator/blob/672ef32de7a1f719c3afdd38147277c88a3f78d0/controllers/metal3.io/baremetalhost_controller.go#L783-L788

https://github.com/metal3-io/baremetal-operator/blob/672ef32de7a1f719c3afdd38147277c88a3f78d0/controllers/metal3.io/baremetalhost_controller.go#L1135